Sources/MongoSwift/BSON/BsonDecoder.swift

```diff
@@ -525,4 +557,127 @@ private extension DecodingError {
         let description = "Expected to decode \(expectation) but found \(type(of: reality)) instead."
         return .typeMismatch(expectation, Context(codingPath: path, debugDescription: description))
     }
+
+    static func _numberMismatch(at path: [CodingKey], expectation: Any.Type, reality: BsonValue?) -> DecodingError {
+        let description = "Expected to find a value that can be represented as a \(expectation), " +
+                         "but found value \(String(describing: reality)) of type \(type(of: reality)) instead."
+        return .typeMismatch(expectation, Context(codingPath: path, debugDescription: description))
+
+    }
+}
+
+/// This needs to be in this file to access some fileprivate decoder properties
+extension Document: Decodable {
+
+    public init(from decoder: Decoder) throws {
+        // if it's a BsonDecoder we should just short-circuit and return the container document
+        if let bsonDecoder = decoder as? _BsonDecoder {
+            let topContainer = bsonDecoder.storage.topContainer
+            guard let doc = topContainer as? Document else {
+                throw DecodingError._typeMismatch(at: [], expectation: Document.self, reality: topContainer)
+            }
+            self = doc
+        // Otherwise get a keyed container and decode each key one by one
+        } else {
+            let container = try decoder.container(keyedBy: _BsonKey.self)
+            var output = Document()
+            for key in container.allKeys {
+                let k = key.stringValue
+                output[k] = try Document.recursivelyDecodeKeyed(key: key, container: container)
+            }
+            self = output
+        }
+    }
+
+    /// Switch through all possible BSON types (a document can contain any) and try recursively decoding the value
```

Comment:
does it make sense to actually maintain a list of BSON types, and to iterate through those programmatically for the following guarded let statements? (e.g. `[Binary.self, Double.self, String.self...]`) This would have the benefit of making it very clear which types directly map to the spec, and also seem to cut down on boilerplate. @

Reply:
:+1: